### DISCUSSION TITLE
Different levels of resource filtering

### DISCUSSION THREAD

**Discussion by TorutheRedFox** [LBPUnion/ProjectLighthouse#464] [Ideas]
At the moment, resources are either filtered or not, but a good feature would be to have multiple levels of filtration, ranging from protecting against potentially dangerous resources (i.e. scripts) to only allowing resources that the retail game can generate on its own

**Comment by m88youngling** [comment]
I think I heard one of the devs talking about this the other day. So, are you describing how scripts could be hidden inside a level's resources as normal data and this needs to be patched, or is this a separate topic?

**Reply by TorutheRedFox** [reply]
nah it's basically choosing between allowing mods regardless, allowing mods that don't include potentially harmful content such as weirdly large files, scripts, etc., or completely ban mods altogether (at least ones that upload resources that a retail version of the game wouldn't generate itself, i.e. models, plans that are tagged in a way that wouldn't normally allow you to add them to a bubble, etc.)

**Reply by m88youngling** [reply]
I'm not exactly sure how the present mod filter works, @Slendy or @Zaprit might know more. I know that Jvyden, before leaving the project, told us the following information on his thoughts on mod filtering:

![unknown-200](https://user-images.githubusercontent.com/89745384/187731070-a74efd25-6a70-4881-8a91-6028b9126ddf.png)

I will say that I actually disagree with him about this topic because it's very "end all be all" on mods. I think right now in the stage we are in, our recommendation is that if you're going to run mods, keep your instance private. Beacon, the public instance we're working on, should have total mod filter enabled for security reasons until we can later figure out how to better filter for dangerous mods.

Jvyden's view on it is correct, but leaves no room for further discovery. I have no doubt that in the future we'll figure out how to better filter mods. Still, any potential for exploitation in the current system should still be discussed to protect servers

**Reply by TorutheRedFox** [reply]
ok so here's my view on common resource types

RMesh: allow it after screening the file and its dependencies for any malicious content/exploits
RScript: unless you can decompile and analyze them, no
RGfxMaterial: no because of zero cross-compatibility between games, unless you want to add a mechanism for limiting compatibility to a single game on the server side
RTexture: what harm could that cause? it's basically sent straight to the GPU after decompression
anything else: screen it for possible exploits and allow it if its dependencies also pass


as for dependencies, keep a log of a file and delete it if a dependency for it comes in that gets rejected

**Reply by TorutheRedFox** [reply]
there's also a few resource types that make zero sense to be uploaded, like RPixelShader, RVertexShader, RSPUElf, etc.

**Reply by Slendy** [reply]
As far as I know the game never uploads a mesh file naturally. It has to be modded in somehow so it doesn’t really have a place besides on instances wanting modded content. I already added functionality for reading dependencies of resources and it should be trivial for someone to fork Lighthouse and adjust the resource filtering. Also, can’t you already upload texture files with no issues?

**Reply by TorutheRedFox** [reply]
I'm talking about modded content for a less strict content filtering mode so the instance is still (at least somewhat) secure from resource-related ingame exploits but still allows for modding, meanwhile the more strict mode would just allow resources that the game itself would generate under normal circumstances to be uploaded (as in no security, lax security allowing for mods, or strict security that enforces vanilla content)

**Reply by m88youngling** [reply]
> ok so here's my view on common resource types
> 
> RMesh: allow it after screening the file and its dependencies for any malicious content/exploits RScript: unless you can decompile and analyze them, no RGfxMaterial: no because of zero cross-compatibility between games, unless you want to add a mechanism for limiting compatibility to a single game on the server side RTexture: what harm could that cause? it's basically sent straight to the GPU after decompression anything else: screen it for possible exploits and allow it if its dependencies also pass
> 
> as for dependencies, keep a log of a file and delete it if a dependency for it comes in that gets rejected

This is some fascinating stuff, I've never seen these categorized like that. I don't know a whole lot about it, so it's great to see your perspective
